File: src/main/java/com/machiav3lli/backup/schedules/HandleScheduledBackups.java
package com.machiav3lli.backup.schedules;

import android.content.Context;
import android.content.SharedPreferences;
import android.database.sqlite.SQLiteDatabase;
import android.os.PowerManager;
import android.preference.PreferenceManager;
import android.util.Log;

import com.machiav3lli.backup.items.AppInfo;
import com.machiav3lli.backup.handler.AppInfoHelper;
import com.machiav3lli.backup.handler.BackupRestoreHelper;
import com.machiav3lli.backup.Constants;
import com.machiav3lli.backup.tasks.Crypto;
import com.machiav3lli.backup.handler.FileCreationHelper;
import com.machiav3lli.backup.handler.FileReaderWriter;
import com.machiav3lli.backup.activities.MainActivityX;
import com.machiav3lli.backup.R;
import com.machiav3lli.backup.handler.ShellCommands;
import com.machiav3lli.backup.handler.NotificationHelper;

import java.io.File;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class HandleScheduledBackups {
    static final String TAG = Constants.TAG;

    Context context;
    PowerManager powerManager;
    ShellCommands shellCommands;
    SharedPreferences prefs;
    File backupDir;
    List<BackupRestoreHelper.OnBackupRestoreListener> listeners;

    public HandleScheduledBackups(Context context) {
        this.context = context;
        prefs = PreferenceManager.getDefaultSharedPreferences(context);
        shellCommands = new ShellCommands(prefs, context.getFilesDir());
        powerManager = (PowerManager) context.getSystemService(Context.POWER_SERVICE);
        listeners = new ArrayList<>();
    }

    public void setOnBackupListener(BackupRestoreHelper.OnBackupRestoreListener listener) {
        listeners.add(listener);
    }

    public void initiateBackup(final int id, final int mode, final int subMode, final boolean excludeSystem) {
        new Thread(() -> {
            String backupDirPath = prefs.getString(
                    Constants.PREFS_PATH_BACKUP_DIRECTORY,
                    FileCreationHelper.getDefaultBackupDirPath());
            assert backupDirPath != null;
            backupDir = new File(backupDirPath);
            ArrayList<AppInfo> list = AppInfoHelper.getPackageInfo(
                    context, backupDir, false, prefs.getBoolean(
                            Constants.PREFS_ENABLESPECIALBACKUPS, true));
            ArrayList<AppInfo> listToBackUp;
            switch (mode) {
                case 0:
                    // all apps
                    Collections.sort(list);
                    backup(list, subMode);
                    break;
                case 1:
                    // user apps
                    listToBackUp = new ArrayList<>();
                    for (AppInfo appInfo : list) {
                        if (!appInfo.isSystem()) {
                            listToBackUp.add(appInfo);
                        }
                    }
                    Collections.sort(listToBackUp);
                    backup(listToBackUp, subMode);
                    break;
                case 2:
                    // system apps
                    listToBackUp = new ArrayList<>();
                    for (AppInfo appInfo : list) {
                        if (appInfo.isSystem()) {
                            listToBackUp.add(appInfo);
                        }
                    }
                    Collections.sort(listToBackUp);
                    backup(listToBackUp, subMode);
                    break;
                case 3:
                    // new and updated apps
                    listToBackUp = new ArrayList<>();
                    for (AppInfo appInfo : list) {
                        if ((!excludeSystem || !appInfo.isSystem()) && (appInfo.getLogInfo() == null || (appInfo.getVersionCode() > appInfo.getLogInfo().getVersionCode()))) {
                            listToBackUp.add(appInfo);
                        }
                    }
                    Collections.sort(listToBackUp);
                    backup(listToBackUp, subMode);
                    break;
                case 4:
                    // custom package list
                    listToBackUp = new ArrayList<>();
                    FileReaderWriter frw = new FileReaderWriter(
                            prefs.getString(Constants.PREFS_PATH_BACKUP_DIRECTORY,
                                    FileCreationHelper.defaultBackupDirPath),
                            SchedulerActivity.SCHEDULECUSTOMLIST + id);
                    for (AppInfo appInfo : list) {
                        if (frw.contains(appInfo.getPackageName())) {
                            listToBackUp.add(appInfo);
                        }
                    }
                    Collections.sort(listToBackUp);
                    backup(listToBackUp, subMode);
                    break;
            }
        }).start();
    }

    public void backup(final ArrayList<AppInfo> backupList, final int subMode) {
        if (backupDir != null) {
            new Thread(() -> {
                Crypto crypto = null;
                if (prefs.getBoolean(Constants.PREFS_ENABLECRYPTO,
                        false) && Crypto.isAvailable(context)) {
                    final String userIds = prefs.getString(
                            "cryptoUserIds", "");
                    final String provider = prefs.getString(
                            "openpgpProviderList", "org.sufficientlysecure.keychain");
                    assert userIds != null;
                    crypto = new Crypto(userIds, provider);
                    crypto.bind(context);
                }
                PowerManager.WakeLock wl = powerManager.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, TAG);
                if (prefs.getBoolean("acquireWakelock", true)) {
                    wl.acquire(10 * 60 * 1000L /*10 minutes*/);
                    Log.i(TAG, "wakelock acquired");
                }
                int id = (int) System.currentTimeMillis();
                int total = backupList.size();
                int i = 1;
                boolean errorFlag = false;
                BlacklistsDBHelper blacklistsDBHelper =
                        new BlacklistsDBHelper(context);
                SQLiteDatabase db = blacklistsDBHelper.getReadableDatabase();
                List<String> blacklistedPackages = blacklistsDBHelper
                        .getBlacklistedPackages(db, SchedulerActivity.GLOBALBLACKLISTID);
                for (AppInfo appInfo : backupList) {
                    if (blacklistedPackages.contains(appInfo.getPackageName())) {
                        Log.i(TAG, String.format("%s ignored",
                                appInfo.getPackageName()));
                        i++;
                        continue;
                    }
                    String title = context.getString(R.string.backupProgress) + " (" + i + "/" + total + ")";
                    NotificationHelper.showNotification(context, MainActivityX.class, id, title, appInfo.getLabel(), false);
                    final BackupRestoreHelper backupRestoreHelper = new BackupRestoreHelper();
                    int ret = backupRestoreHelper.backup(context, backupDir, appInfo, shellCommands, subMode);
                    if (ret != 0)
                        errorFlag = true;
                    else if (crypto != null) {
                        crypto.encryptFromAppInfo(context, backupDir, appInfo, subMode, prefs);
                        if (crypto.isErrorSet()) {
                            Crypto.cleanUpEncryptedFiles(new File(backupDir, appInfo.getPackageName()), appInfo.getSourceDir(), appInfo.getDataDir(), subMode, prefs.getBoolean("backupExternalFiles", false));
                            errorFlag = true;
                        }
                    }
                    if (i == total) {
                        String notificationTitle = errorFlag ? context.getString(R.string.batchFailure) : context.getString(R.string.batchSuccess);
                        String notificationMessage = context.getString(R.string.sched_notificationMessage);
                        NotificationHelper.showNotification(context, MainActivityX.class, id, notificationTitle, notificationMessage, true);
                    }
                    i++;
                }
                if (wl.isHeld()) {
                    wl.release();
                    Log.i(TAG, "wakelock released");
                }
                for (BackupRestoreHelper.OnBackupRestoreListener l : listeners)
                    l.onBackupRestoreDone();
                blacklistsDBHelper.close();
            }).start();
        }
    }
}

File: app/src/main/java/com/machiav3lli/backup/schedules/HandleScheduledBackups.kt
/*
 * OAndBackupX: open-source apps backup and restore app.
 * Copyright (C) 2020  Antonios Hazim
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.machiav3lli.backup.schedules

import android.content.Context
import android.util.Log
import androidx.work.Data
import androidx.work.OneTimeWorkRequest
import androidx.work.WorkManager
import com.machiav3lli.backup.Constants.classTag
import com.machiav3lli.backup.R
import com.machiav3lli.backup.activities.MainActivityX
import com.machiav3lli.backup.activities.SchedulerActivityX
import com.machiav3lli.backup.dbs.BlacklistDatabase
import com.machiav3lli.backup.dbs.Schedule
import com.machiav3lli.backup.handler.BackendController
import com.machiav3lli.backup.handler.BackupRestoreHelper
import com.machiav3lli.backup.handler.NotificationHelper
import com.machiav3lli.backup.items.AppInfo
import com.machiav3lli.backup.tasks.ScheduledWork
import com.machiav3lli.backup.utils.FileUtils.BackupLocationIsAccessibleException
import com.machiav3lli.backup.utils.LogUtils
import com.machiav3lli.backup.utils.StorageLocationNotConfiguredException
import com.machiav3lli.backup.utils.checkStoragePermissions

class HandleScheduledBackups(private val context: Context) {
    private val listeners: MutableList<BackupRestoreHelper.OnBackupRestoreListener> = mutableListOf()

    fun setOnBackupListener(listener: BackupRestoreHelper.OnBackupRestoreListener) {
        listeners.add(listener)
    }

    fun initiateBackup(id: Int, mode: Schedule.Mode?, subMode: Schedule.SubMode?, excludeSystem: Boolean,
                       customList: Set<String>) {
        Thread {
            Runnable {
                val notificationId = System.currentTimeMillis().toInt()
                NotificationHelper.showNotification(context, MainActivityX::class.java, notificationId,
                        String.format(context.getString(R.string.fetching_action_list),
                                context.getString(R.string.backup)), "", true)
                val list: List<AppInfo>
                list = try {
                    BackendController.getApplicationList(context)
                } catch (e: BackupLocationIsAccessibleException) {
                    Log.e(TAG, "Scheduled backup failed due to ${e.javaClass.simpleName}: $e")
                    LogUtils.logErrors(context, e.toString())
                    return@Runnable
                } catch (e: StorageLocationNotConfiguredException) {
                    Log.e(TAG, "Scheduled backup failed due to ${e.javaClass.simpleName}: $e")
                    LogUtils.logErrors(context, e.toString())
                    return@Runnable
                }
                val inCustomList = { packageName: String -> customList.isEmpty() || customList.contains(packageName) }
                val predicate: (AppInfo) -> Boolean = when (mode) {
                    Schedule.Mode.USER -> { appInfo: AppInfo ->
                        appInfo.isInstalled && !appInfo.isSystem && inCustomList(appInfo.packageName)
                    }
                    Schedule.Mode.SYSTEM -> { appInfo: AppInfo ->
                        appInfo.isInstalled && appInfo.isSystem && inCustomList(appInfo.packageName)
                    }
                    Schedule.Mode.NEW_UPDATED -> { appInfo: AppInfo ->
                        (appInfo.isInstalled && (!excludeSystem || !appInfo.isSystem)
                                && (!appInfo.hasBackups || appInfo.isUpdated)
                                && inCustomList(appInfo.packageName))
                    }
                    else -> { appInfo: AppInfo -> inCustomList(appInfo.packageName) }
                }
                val listToBackUp = list
                        .filter(predicate)
                        .toList()
                        .sortedWith { m1: AppInfo, m2: AppInfo ->
                            m1.packageLabel.compareTo(m2.packageLabel, ignoreCase = true)
                        }
                val selectedList = listToBackUp.map { it.packageName }
                startScheduledBackup(id, selectedList, subMode?.value
                        ?: Schedule.SubMode.BOTH.value, notificationId)
            }.run()
        }.start()
    }

    private fun startScheduledBackup(id: Int, selectedList: List<String>, subMode: Int, notificationId: Int) {
        if (checkStoragePermissions(context)) {
            val blacklistDao = BlacklistDatabase.getInstance(context).blacklistDao
            val globalBlacklist = blacklistDao.getBlacklistedPackages(SchedulerActivityX.GLOBAL_ID)
            val customBlacklist = blacklistDao.getBlacklistedPackages(id)
            val blacklistedPackages = globalBlacklist.plus(customBlacklist).toSet()
            val data: Data = Data.Builder()
                    .putStringArray("blackList", blacklistedPackages.toTypedArray())
                    .putStringArray("selectedPackages", selectedList.toTypedArray())
                    .putInt("selectedMode", subMode)
                    .putInt("notificationId", notificationId)
                    .build()

            val oneTimeWorkRequest = OneTimeWorkRequest.Builder(ScheduledWork::class.java)
                    .setInputData(data)
                    .build()

            WorkManager.getInstance(context)
                    .enqueue(oneTimeWorkRequest)
        } else {
            val notificationTitle = context.getString(R.string.sched_notificationMessage)
            val notificationMessage = context.getString(R.string.permission_not_granted)
            NotificationHelper.showNotification(context, MainActivityX::class.java, notificationId,
                    notificationTitle, notificationMessage, true)
        }
    }

    companion object {
        private val TAG = classTag(".HandleScheduledBackups")
    }
}